Camel case binding runtime fixes

Benjamin-Dobell opened the pull request:
There's a few commits here, but the common thread is fixing runtime bugs in the recently introduced camel-case bindings.

# godot.lib.api

This PR introduces a `godot.lib.api` module. This module allows you to write camel-case binding insensitive code.

The top-level export is a proxy wrapping the `godot` API. When you consume APIs (access properties or call methods), the resultant values are also wrapped in proxies. The purpose of these proxies are to (as seamlessly as possible) translate non-camel-case APIs to camel-case APIs _if_ the project has camel-case bindings enabled.

The primary purpose of the module is for GodotJS internals to consume this API. A benefit of this approach is that the codegen code for example is now much cleaner. It now much more closely resembles the original implementation prior to https://github.com/godotjs/GodotJS/pull/65 being merged i.e. the implementation for the most part no longer needs to concern itself with whether camel-case bindings are enabled, it's abstracted away. However, beyond cleaning up code, there were some bugs I'd missed with the use of flags e.g. `Object.ConnectFlags.CONNECT_ONE_SHOT`. This was coming through as undefined when camel-case bindings were enabled, because the correct property name when camel-case bindings are enabled is `Object.ConnectFlags.ConnectOneShot`. `godot.lib.api` now seamlessly translates `Object.ConnectFlags.CONNECT_ONE_SHOT` into `Object.ConnectFlags.ConnectOneShot`.

The proxies do of course introduce overhead, however it ought to be negligible for user code, since the proxies only come into play for `godot.lib.api`, there's no overhead when consuming `godot` as per normal. The main place where some overhead may have appeared is in the editor, say for example during codegen. Codegen does make fairly extensive use of `godot.lib.api`, because the codegen code needs to operate irrespective of the user's camel-case bindings setting.

Somewhat inadvertently, `godot.lib.api` is also useful if users wish to distribute libraries that are themselves camel-case binding agnostic. In their library it ought to be as simple as replacing an import of `godot` with `godot.lib.api` and their library will be safe to import in projects that have camel-case bindings enabled. 

# AMD module return value support

I've implemented support for AMD return values, whilst maintaining support for the `exports` dependency (CommonJS-style). This can be achieved in TypeScript by taking advantage of TypeScript's [`export =` syntax](https://www.typescriptlang.org/docs/handbook/modules/reference.html#export--and-import--require). This was necessary in order to export a proxy rather than an object from a module, not as the default value, but rather as the module export itself. This allows `godot.lib.api` to be a drop-in replacement for `godot`.

# jsb.runtime entry-point is now main.ts

`jsb.inject.ts` was the previous entry-point.

When generating an AMD bundle, the TypeScript compiler bundles all ES6/ESM modules by wrapping them in AMD's `define()`. Any TS files which are not ES modules are included directly in the bundle (side effects executed on load). Consequently `jsb.inject.ts`, which was previously the entry-point was not implemented as an ES module, it used `require` syntax exclusively. However, this meant that the implementation was not able to import any types, since even the introduction of `import type` syntax (no runtime imports) caused tsc to treat the file as an ES module.

Thus, `jsb.inject.ts` was not type-safe. To make the implementation less error prone, I've added (imported) types into `jsb.inject.ts`. Which, as mentioned, caused tsc to treat this file as an ES module i.e. no longer an entry point. To rectify this, I've introduced `main.ts`, which is itself not an ES module, it simply requires `jsb.inject.ts`.

Related to this, [the final commit](https://github.com/godotjs/GodotJS/pull/74/commits/a06e2c0a3f5536d302b329db7ef5d5649fde225b) in this PR is a largely non-functional change. The AMD module loader has logic to set `require.main`, and complained (logged) when it was unable to do so. Because we're now calling `require` on start-up of our runtime bundle, this logging was appearing on start-up. The final change-set simply ensures this logging doesn't take place, since it's expected `require.main` is unavailable at this point.

# [fix: TypeLoader _post_bind_ class loading](https://github.com/godotjs/GodotJS/pull/74/commits/59bb8d0ea10a6c2655500e234f2e0dbc6dddb972)

i.e. Fixed typeloader callbacks being unable to load types.

Previously, when calling into `_post_bind_` (user code) we had in scope a `NativeClassInfoPtr` (`internal::SArray<NativeClassInfo, NativeClassID>::Pointer`) which holds a lock on the environment's `native_classes_` array. This resulted in crash if the callback itself attempted to load another type.

# [fix: QuickJS support and fix SceneNode codegen.](https://github.com/godotjs/GodotJS/pull/74/commits/3179623506b4a435f0a5a1a4d5c3af3aeb6eeaef)

This commit is really a follow-up (fix-up commit) for the first commit in the series. Typically I squash these back into the original commit, but it includes a decent commit message which will make a `git blame` on these lines of code much more informative. So I decided to keep it separate.

It's worth noting that this commit moves the `[[FloatType]]` and `[[IntegerType]]` symbols from TypeScript into C++, now defined in the same location as all our other symbols. The motivation for this was that I've introduced a new symbol `[[ProxyTarget]]` which I wanted to access from both C++ and TypeScript. Rather than having 3 different ways in which symbols are defined, I've consolidated. Accessing the `[[ProxyTarget]]` property in C++ became necessary when I realised our `QuickJS` v8 interop layer doesn't implement `IsProxy()`, and `GetTarget()`, and that they're in fact impossible to implement (without changing QuickJS itself). Thus the `[[ProxyTarget]]` symbol is used in place of `GetTarget()`. 

# [fix: Native object signals not firing in camel-case bindings mode](https://github.com/godotjs/GodotJS/pull/74/commits/5073bc4fcc9ac6685a9eead258d751e6cd50b135)
The LRU StringName cache was implemented in parallel to camel-case bindings. It made changes to the way GodotJS' pseudo-`Signal`s operate. Essentially the JS property name was assumed to be the same as the Signal name in Godot/GDScript. However, with the introduction of camel-case bindings project setting, camel-case naming conventions are used for all properties, including signals. So when both camel-case bindings and the LRU cache landed together on main, signals stopped working for projects using camel-case API bindings.

Benjamin-Dobell commented:
[Force push](https://github.com/godotjs/GodotJS/compare/7821698de50289213754046c24489bb9abf2b467..a06e2c0a3f5536d302b329db7ef5d5649fde225b) removed changes to `package.json` files I didn't mean to include in the PR. I use PNPM and it keeps re-injecting this change itself 😅

Benjamin-Dobell commented:
Oh, I forgot to mention. This PR also includes codegen fixes. Our default tsconfig has:

```
skipLibCheck": false
```

However, if you swap that to true in any project using GodotJS, you'll see that the codegen files we were generating were actually invalid.

The most common error was including defaults, which is not permitted on type declarations, only implementations (reverse behavior compared to C++!) However, the default values are useful from a documentation perspective, so I've kept them, but they're now a comment.

There were other errors too, which I imagine I was responsible for introducing in my recent codegen changes.

ialex32x commented on bridge/jsb_type_convert.cpp:
```diff
@@ -141,6 +141,20 @@ namespace jsb
+#endif
+        {
+            v8::Local<v8::Object> object = v8::Local<v8::Object>::Cast(p_jval);
+            v8::MaybeLocal<v8::Value> target = object->Get(context, Environment::wrap(isolate)->get_symbol(Symbols::ProxyTarget));
```
It may introduce too much overhead on the performance. 
It's better to keep the C++ transpiler simple and leave these responsibilities in the script layer? For example, do `some_gdobj.some_call( internal.unproxy( proxy_obj ) )`?

Benjamin-Dobell replied:
Initially I was using v8's `IsProxy()` and `GetTarget()` APIs which would have been much more efficient but unfortunately there's no equivalent for QuickJS.

However, I'm not sure this PR can reliably function without this logic existing in C++, ecause we only want to unwrap proxies on calls to natively implemented functions. The proxy handler doesn't know which functions are implemented natively and which functions are implemented in TS.

That said, it's entirely possible this PR is just fundamentally the wrong approach.

The main issue I was wanting to address was that GodotJS internals must function in either Godot API binding mode.

In my previous PR I'd made the codegen function in either binding mode by wrapping every potentially problematic property access with an explicit function call that was aware of the camel case binding mode, and then in TS converted from snake-case to camel-case. This approach is manual and error-prone since it requires awareness of which properties are affected by camel-case bindings; it's very easy to accidentally access a property manually instead of using the wrapper function. I had missed some such cases, resulting in `undefined` properties coming through in unexpected places.

Ideally, I'd like to abstract away camel-case binding awareness entirely, so code is easier to implement/maintain, which is what I've tried to achieve in this PR.

An alternative approach with zero runtime overhead would be to provide separate implementations of GodotJS' internals for camel-case bindings i.e. the JS file loaded would be different depending on whether the camel-case bindings setting is enabled or not. The problem then becomes maintaining the two different runtimes. Presumably we wouldn't actually want to maintain two separate lots of TS files, because they could come out of sync. So we could try generate them from some intermediary representation. Buuuuuut, that's a lot of changes, and I'm trying (and admittedly failing) to keep the camel-case bindings stuff out of your way, and minimise maintenance burden.

Honestly, the simplest solution from my perspective would be just to swap to the camel-case binding scheme, as the only API binding implementation 😅 But I can understand why you may not want to do that.

ialex32x replied:
The `p_jval->IsProxy()` and `GetProperty(symbol)` checks in type_convert will be performed on every call where the object wrap/unwrap is needed, which has a more critical impact on performance. 

Never mind, it's in an early stage for essential features and possibilities instead of restrictions. Let's merge and improve it later.

ialex32x commented on bridge/jsb_type_convert.cpp:
```diff
@@ -471,9 +485,22 @@ namespace jsb
+            if (p_jval->IsProxy())
+#endif
+            {
+                v8::MaybeLocal<v8::Value> target = self->Get(context, Environment::wrap(isolate)->get_symbol(Symbols::ProxyTarget));
```
It may introduce too much overhead on the performance. 
It's better to keep the C++ transpiler simple and leave these responsibilities in the script layer? For example, do `some_gdobj.some_call( internal.unproxy( proxy_obj ) )`?

ialex32x commented:
By the way, in the last PR which introduced the auto naming conversion, 
`List<StringName> exposed_class_list = internal::NamingUtil::get_exposed_class_list();` will instantiate a complete classname list immediately on startup even if the naming conversion is disabled.
Is it possible to make it totally omitted and have no additional side effects on the startup time?

Benjamin-Dobell commented on bridge/jsb_environment.cpp:
```diff
@@ -1305,18 +1341,20 @@ namespace jsb
-            NativeClassInfoPtr class_info = class_register->register_func(ClassRegister {
+            // We can't hold NativeClassInfoPtr here because on_class_post_bind calls into user code, which may wish to load
+            // additional classes and thus needs the lock released.
+            const NativeClassInfo* class_info = (NativeClassInfo*)class_register->register_func(ClassRegister {
```
Thanks for catching this! Real brain fade moment.

I initially implemented it the way I have now in the [commit I've just pushed](https://github.com/godotjs/GodotJS/pull/74/commits/ca0e3e2886fe0c27488ef8e0f7cf70c6febf6b11). However, I refactored and stupidly tried to optimize the solution in a way that made no sense at all 🤦

Benjamin-Dobell replied:
Please let me know if the latest approach is problematic.

Benjamin-Dobell commented:
> By the way, in the last PR which introduced the auto naming conversion, `List<StringName> exposed_class_list = internal::NamingUtil::get_exposed_class_list();` will instantiate a complete classname list immediately on startup even if the naming conversion is disabled. Is it possible to make it totally omitted and have no additional side effects on the startup time?

Definitely open to alternative suggestions, however the goal for the approach taken was to achieve as close to zero runtime overhead with camel-case bindings enabled vs not enabled as possible, for user's game code anyway (I'm less concerned about Editor performance).

GodotJS was already looking up replacement type names; when camel-case bindings are enabled that replacement map is just filled with a lot of replacements. Admittedly, hash map lookups aren't O(1), so there is some overhead, but it's minimal. There's also a tiny hit to startup time as the map is populated.

An alternative approach would be to convert type names on the fly each time one is encountered. Doing so every time would be lot of overhead. They could be cached though, so the performance hit would be the first time the type is referenced. That's shifting the performance impact from startup to general runtime _but_ it's pretty negligible. So I'd be happy to move to this approach if you prefer.

I've done zero profiling on this, but I'd (perhaps naively) assumed the memory usage of of this map would be negligible compared to how much data is already in memory about these types for scripting purposes (class names, method names, parameters, types etc.) Granted, I implemented this before (or in parallel with) the work you'd done on the LRU cache. Now that I've seen that work, I assume you made that change for a good reason, so I'm definitely second guessing my initial assumption!

How are you going about profiling thus far? If you've any tips/workflows/tests you're able to share that'd be most appreciated. Thus far I'm only building against v8; I haven't even tried QuickJS or web yet 😅

Benjamin-Dobell commented:
Oh, sorry, I just re-read.

> even if the naming conversion is disabled

You're quite right. It's entirely possible to eliminate the creation of this class list on startup if camel-case bindings are disabled. This class list is used in two places:

- Startup, for remapping class names. However, this need not occur if camel-case bindings are enabled.
- Codegen. The entire class list is needed for codegen, but this is editor only, so could be deferred.

I'll take a crack at eliminating this start-up overhead if camel-case bindings are disabled _and_ it's a non-editor build.

Benjamin-Dobell commented:
-  > By the way, in the last PR which introduced the auto naming conversion, `List<StringName> exposed_class_list = internal::NamingUtil::get_exposed_class_list();` will instantiate a complete classname list immediately on startup even if the naming conversion is disabled. Is it possible to make it totally omitted and have no additional side effects on the startup time?
  
    Addressed in 914f3b5fa7a7b6f85f169a8148e2a28f77f85558
- Included two other codegen fixes for issues I encountered when testing codegen with the GodotJSExamples project.
- Rebased

ialex32x commented:
> > By the way, in the last PR which introduced the auto naming conversion, `List<StringName> exposed_class_list = internal::NamingUtil::get_exposed_class_list();` will instantiate a complete classname list immediately on startup even if the naming conversion is disabled. Is it possible to make it totally omitted and have no additional side effects on the startup time?
> 
> Definitely open to alternative suggestions, however the goal for the approach taken was to achieve as close to zero runtime overhead with camel-case bindings enabled vs not enabled as possible, for user's game code anyway (I'm less concerned about Editor performance).
> 
> GodotJS was already looking up replacement type names; when camel-case bindings are enabled that replacement map is just filled with a lot of replacements. Admittedly, hash map lookups aren't guaranteed O(1), so there is perhaps rarely some overhead, but it's unlikely/minimal. There's also a tiny hit to startup time as the map is populated.
> 
> An alternative approach would be to convert type names on the fly each time one is encountered. Doing so every time would be lot of overhead. They could be cached though, so the performance hit would be the first time the type is referenced. That's shifting the performance impact from startup to general runtime _but_ it's pretty negligible. So I'd be happy to move to this approach if you prefer.
> 
> I've done zero profiling on this, but I'd (perhaps naively) assumed the memory usage of of this map would be negligible compared to how much data is already in memory about these types for scripting purposes (class names, method names, parameters, types etc.) Granted, I implemented this before (or in parallel with) the work you'd done on the LRU cache. Now that I've seen that work, I assume you made that change for a good reason, so I'm definitely second guessing my initial assumption!
> 
> How are you going about profiling thus far? If you've any tips/workflows/tests you're able to share that'd be most appreciated. Thus far I'm only building against v8; I haven't even tried QuickJS or web yet 😅

There isn't an existing workflow for profiling yet (will be considered later). Just a concern because the creation of the Environment class is relatively heavy. More optimizations will be considered in future versions (after all planned features are relatively finished). But anyway, it's reasonable to initialize it early for better runtime performance. I appreciate all of your PRs. Thanks a lot.

Benjamin-Dobell commented:
Thanks! I really appreciate the taking the time to review and merge all these PRs 🙏

> Can we connect through Discord/TG? Mail me if it's OK😃

Absolutely, I've just sent you an email with my Discord username.
